ktor-server/ktor-server-plugins/ktor-server-di/jvm/test/io/ktor/server/plugins/di/utils/TypesTest.kt

```diff
@@ -20,24 +20,40 @@ class TypesTest {
 
     @Test
     fun supertypes() {
-        val supertypes = typeInfo<Cat>().hierarchy()
+        val supertypes = typeInfo<Cat>().hierarchy().toSet()
         assertContains(supertypes, typeInfo<Animal>())
         assertContains(supertypes, typeInfo<Pet>())
         assertContains(supertypes, typeInfo<Companion>())
         assertFalse("Should not contain Any") { typeInfo<Any>() in supertypes }
     }
 
     @Test
-    fun parameterizedSupertypes() {
-        val supertypes = typeInfo<ArrayList<Cat>>().hierarchy()
+    fun `parameterized supertypes`() {
+        val supertypes = typeInfo<ArrayList<Cat>>().hierarchy().toSet()
         assertContains(supertypes, typeInfo<List<Cat>>())
         assertContains(supertypes, typeInfo<Collection<Cat>>())
     }
 
     @Test
-    fun multiParameterizedSupertypes() {
-        val supertypes = typeInfo<LinkedHashMap<String, Cat>>().hierarchy()
+    fun `multi parameterized supertypes`() {
+        val supertypes = typeInfo<LinkedHashMap<String, Cat>>().hierarchy().toSet()
         assertContains(supertypes, typeInfo<HashMap<String, Cat>>())
         assertContains(supertypes, typeInfo<Map<String, Cat>>())
     }
+
+    @Test
+    fun `parameterized arg supertypes`() {
+        val supertypes = typeInfo<List<Cat>>().typeParametersHierarchy().toSet()
+        assertContains(supertypes, typeInfo<List<Pet>>())
+        assertContains(supertypes, typeInfo<List<Animal>>())
+        assertContains(supertypes, typeInfo<List<Companion>>())
```

Comment:
I think we should also test ordering as we rely on it in `OutTypeArgumentsSupertypes`. So it is better to use `assertContentEquals` to test both elements and their order.

```suggestion
        val actualSupertypes = typeInfo<List<Cat>>().typeParametersHierarchy()
        val expectedSupertypes = sequenceOf(
            typeInfo<List<Animal>>(),
            typeInfo<List<Pet>>(),
            typeInfo<List<Companion>>(),
        )

        assertContentEquals(expectedSupertypes, actualSupertypes)
```

Reply:
Btw my changed code won't pass the test because it lacks `typeInfo<List<Cat>>()` at the fist position. Should it be included?

Reply:
Oh yeah that makes sense.

Reply:
I think I'll drop the `mapInclude` function and ensure that all covariant mapping functions include the input key.  That should simplify things a little.

Reply:
Not that this would affect this function... I guess it makes more sense to only include supertypes.

Reply:
Changed my mind again, the type parameters supertypes always includes the input because it's iterating over all combinations, so I ought to do the same with the hierarchy call 😄